azurelinuxagent/common/protocol/imds.py

```diff
@@ -242,15 +243,46 @@ def __init__(self, version=APIVERSION):
             'User-Agent': restutil.HTTP_USER_AGENT_HEALTH,
             'Metadata': True,
         }
-        pass
+        self._regex_imds_ioerror = re.compile(r".*HTTP Failed. GET http://[^ ]+ -- IOError timed out -- [0-9]+ attempts made")
+        self.protocol_util = get_protocol_util()
 
-    @property
-    def compute_url(self):
-        return BASE_URI.format(IMDS_ENDPOINT, 'compute', self._api_version)
+    def _get_metadata_url(self, endpoint, resource_path):
+        return BASE_URI.format(endpoint, resource_path, self._api_version)
 
-    @property
-    def instance_url(self):
-        return BASE_URI.format(IMDS_ENDPOINT, '', self._api_version)
+    def _http_get(self, endpoint, resource_path, headers):
+        url = self._get_metadata_url(endpoint, resource_path)
+        return restutil.http_get(url, headers=headers, use_proxy=False)
+
+    def get_metadata(self, resource_path, is_health):
+        """
+        Get metadata from IMDS, falling back to Wireserver endpoint if necessary.
+
+        :param str resource_path: path of IMDS resource
+        :param bool is_health: True if for health/heartbeat, False otherwise
+        :return: Tuple<is_request_success:bool, response:str>
+            is_request_success: True when connection succeeds, False otherwise
+            response: response from IMDS on request success, failure message otherwise
+        """
+        headers = self._health_headers if is_health else self._headers
+        endpoint = IMDS_ENDPOINT
+        try:
+            resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
+        except HttpError as e:
+            logger.warn("Unable to connect to primary IMDS endpoint {0}", endpoint)
+            if not self._regex_imds_ioerror.match(str(e)):
+                raise e
+            endpoint = self.protocol_util.get_wireserver_endpoint()
+            try:
+                resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
+            except HttpError as e:
+                logger.warn("Unable to connect to backup IMDS endpoint {0}", endpoint)
+                if not self._regex_imds_ioerror.match(str(e)):
+                    raise e
+                return False, "IMDS error in /metadata/{0}: Unable to connect to endpoint".format(resource_path)
```

Comment:
I don't feel that verbosity would help since the HttpError message is long and only provides "IOError timed out" within it and seems to be vague and not 100% accurate. There was no wait and the return was almost instant when running my tests on a VM.

I try to summarize the connection issue with a (possibly too simple) "Unable to connect to endpoint" message. I can change this as desired but I don't think the HttpError message adds value.

Reply:
If we need to debug an error here, is there anything useful we can add on top of "Unable to connect to endpoint"?

Reply:
This currently captures everything IMDS team needs for troubleshooting. Please provide an example of what you have an mind.

Any additional info that can potentially help with root cause analysis is always welcome and appreciated.